src/modules/settings/components/Table.jsx

```diff
@@ -0,0 +1,242 @@
+      }
+    }
+    document.addEventListener('mousedown', handleClickOutside);
+  }, [ref]);
+}
+
+const EditCell = ({rowData, dataKey, onChange, onMouseOver, onMouseLeave, onClick, ...props}) => {
+  switch (rowData.status) {
+    case 'HOVERING':
+      return (
+        <Cell
+          {...props}
+          className="table-content-editing"
+          onMouseLeave={() => onMouseLeave && onMouseLeave('hovering', rowData.id)}>
+          <input
+            className="bttv-rs-input"
+            style={{top: 7, height: 30, opacity: 0.5}}
+            defaultValue={rowData[dataKey]}
+            onClick={() => onClick('edit', rowData.id)}
+          />
+        </Cell>
+      );
+
+    case 'EDIT': {
+      const wrapperRef = useRef(null);
+      useOutsideAlerter(wrapperRef, onClick, rowData);
+      return (
+        <Cell {...props} className="table-content-editing">
+          <input
+            ref={wrapperRef}
+            className="bttv-rs-input"
+            style={{top: 7, height: 30}}
+            defaultValue={rowData[dataKey]}
+            onChange={(event) => onChange(rowData.id, dataKey, event.target.value)}
+          />
+        </Cell>
+      );
+    }
+
+    default:
+      return (
+        <Cell
+          {...props}
+          onMouseOver={() => onMouseOver && onMouseOver('hovering', rowData.id)}
+          style={{paddingTop: 13, paddingLeft: 10}}>
+          <span className="table-content-edit-span">{rowData[dataKey]}</span>
+        </Cell>
+      );
+  }
+};
+
+const ActionCell = ({rowData, dataKey, onClick, ...props}) => (
+  <Cell {...props} style={{padding: '6px 0'}}>
+    <Button style={{float: 'right', marginRight: 10}} appearance="link" onClick={() => onClick('delete', rowData.id)}>
+      <Icon icon={minus} />
+    </Button>
+  </Cell>
+);
+
+function EditTable({options, setData, data, ...props}) {
+  const modal = useContext(ModalContext);
+
+  if (!modal.enter) return <Placeholder.Graph active height={data.length * 46} />;
+
+  const handleChange = (id, key, value) => {
+    const nextData = Object.assign([], data);
+    nextData.find((item) => item.id === id)[key] = value;
+    setData(nextData);
+  };
+
+  const handleDeleteState = (id) => {
+    const nextData = Object.assign([], data).filter((item) => item.id !== id);
+    setData(nextData);
+  };
+
+  const handleHoveringState = (id) => {
+    const nextData = Object.assign([], data);
+    const activeItem = nextData.find((item) => item.id === id);
+    activeItem.status = activeItem.status ? null : 'HOVERING';
+    setData(nextData);
+  };
+
+  const handleEditState = (id) => {
+    const nextData = Object.assign([], data);
+    const activeItem = nextData.find((item) => item.id === id);
+    activeItem.status = activeItem.status !== 'EDIT' ? 'EDIT' : null;
+    setData(nextData);
+  };
+
+  const nextId = () => {
+    let newId = 0;
+    for (let i = 0; data.find((item) => item.id === i) !== undefined; i++) {
```

Comment:
this operation is quite slow. assuming an ordered list, wouldn't this be `lastItemId + 1`